### PR TITLE
#712 Fixed BkBasic request handling

### PR DESCRIPTION
**Pull request by rui-castro**
Issue #712 
This PR makes `BkBasic` consume the entire request body before processing a new request.
If this is not done, `BkBasic` will treat the remaining body of a request as a new request.

### CODE REVIEW

**Comment by davvd** [comment]
@rui-castro Thanks, someone will review your pull request soon

**Comment by davvd** [comment]
@essobedo this pull request is for you, please review

File: src/main/java/org/takes/http/BkBasic.java [hunk 1]
```diff
@@ -68,15 +68,15 @@
...
 
-    @SuppressWarnings ("PMD.AvoidInstantiatingObjectsInLoops")
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
     @Override
     public void accept(final Socket socket) throws IOException {
         try (
-            final InputStream input = socket.getInputStream();
-            final BufferedOutputStream output = new BufferedOutputStream(
+            InputStream input = socket.getInputStream();
+            BufferedOutputStream output = new BufferedOutputStream(
```
Comment (essobedo) [inline src/main/java/org/takes/http/BkBasic.java]: @rui-castro Why removing `final`?

File: src/main/java/org/takes/tk/TkConsumeBody.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import org.takes.rq.RqLengthAware;
+
+/**
+ * A {@link Take} decorator that consumes the remaining body after
+ * being processed by the inner {@link Take}.
+ *
+ * <p>The class is immutable and thread-safe.
+ * @author Rui Castro (rui.castro@gmail.com)
+ * @version $Id$
+ * @since 0.43
```
Comment (essobedo) [inline src/main/java/org/takes/tk/TkConsumeBody.java]: @rui-castro Why `0.43`?

Comment (rui-castro) [inline src/main/java/org/takes/tk/TkConsumeBody.java]: @essobedo I don't know what the next version will be, but, since the current version is 0.42, I assumed it will be 0.43 :)

File: src/main/java/org/takes/rq/CapInputStream.java [hunk 3]
```diff
@@ -60,10 +59,7 @@
 
     @Override
     public int available() throws IOException {
-        return (int) Math.min(
-            (long) Integer.MAX_VALUE,
-            Math.max((long) this.origin.available(), this.more)
-        );
+        return (int) Math.min(this.origin.available(), this.more);
```
Comment (essobedo) [inline src/main/java/org/takes/rq/CapInputStream.java]: @rui-castro even if this makes sense to me, this changes the whole initial logic of this method, so I let the architect decide. 
@yegor256 What do you think about this change?

File: src/main/java/org/takes/http/BkBasic.java (hunk 1 above)
Comment (essobedo) [inline src/main/java/org/takes/http/BkBasic.java]: @rui-castro I don't like the idea to mix fixes on a given PR. One PR one fix

File: src/main/java/org/takes/rq/CapInputStream.java (hunk 3 above)
Comment (essobedo) [inline src/main/java/org/takes/rq/CapInputStream.java]: @yegor256 What do you think about this change?

File: src/test/java/org/takes/rq/RqLengthAwareTest.java [hunk 4]
```diff
@@ -62,11 +61,12 @@ public void addsLengthToBody() throws IOException {
...
 
     /**
-     * RqLengthAware can add BIG length to body.
+     * RqLengthAware can respect the real body length over
+     * "Content-Length" value.
      * @throws IOException If some problem inside
      */
     @Test
-    public void addsBigLengthToBody() throws IOException {
+    public void respectRealBodyLength() throws IOException {
```
Comment (essobedo) [inline src/test/java/org/takes/rq/RqLengthAwareTest.java]: @rui-castro it is misspelled, it should rather be `respectsRealBodyLength` with a s

File: src/test/java/org/takes/tk/TkSlf4jTest.java [hunk 5]
```diff
@@ -43,7 +49,7 @@
      */
     @Test
     public void logsMessage() throws IOException {
-        new TkSlf4j(new TkText("test")).act(new RqFake());
+        new TkSlf4j(new TkEmpty()).act(new RqFake());
```
Comment (essobedo) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @rui-castro Why this change?

File: src/test/java/org/takes/tk/TkSlf4jTest.java [hunk 6]
```diff
@@ -55,4 +61,36 @@ public void logsException() throws IOException {
         new TkSlf4j(new TkFailure(new IOException(""))).act(new RqFake());
     }
 
+    /**
+     * {@link TkSlf4j} can output an empty body for {@link TkEmpty}.
+     * @throws IOException if some I/O problem occurred.
+     */
+    @Test
+    public void testEmptyResponseBody() throws IOException {
```
Comment (essobedo) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @rui-castro Please add a test in your unit test

File: src/main/java/org/takes/rq/CapInputStream.java (hunk 3 above)
Comment (rui-castro) [inline src/main/java/org/takes/rq/CapInputStream.java]: @essobedo yes, it changes the behaviour. This class is used in [RqLengthAware](https://github.com/yegor256/takes/blob/master/src/main/java/org/takes/rq/RqLengthAware.java) that is supposed to respect the Content-Length header. Using the max() between available bytes and `more` value, one can read beyond the Content-Length mark.

File: src/test/java/org/takes/rq/CapInputStreamTest.java [hunk 7]
```diff
@@ -69,16 +68,22 @@ public void closesStream() throws Exception {
...
+     * @throws IOException If some problem inside
      */
     @Test
-    public void skipsOnStream() throws Exception {
-        final long skip = 25L;
-        final InputStream stream = Mockito.mock(InputStream.class);
-        final CapInputStream wrapper = new CapInputStream(stream, 50L);
-        wrapper.skip(skip);
-        Mockito.verify(stream, Mockito.times(1)).skip(skip);
+    public void skipsOnStreamAndDecreaseAvailable() throws IOException {
```
Comment (rui-castro) [inline src/test/java/org/takes/rq/CapInputStreamTest.java]: ok

File: src/test/java/org/takes/rq/RqLengthAwareTest.java (hunk 4 above)
Comment (rui-castro) [inline src/test/java/org/takes/rq/RqLengthAwareTest.java]: ok

File: src/test/java/org/takes/tk/TkSlf4jTest.java (hunk 5 above)
Comment (rui-castro) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @essobedo to decrease the data coupling, `TkEmpty` is already used in another test and for this purpose, I think `TkEmpty` serves the same function as `TkText`.

File: src/test/java/org/takes/tk/TkSlf4jTest.java [hunk 8]
```diff
@@ -55,4 +61,36 @@ public void logsException() throws IOException {
...
+        );
+    }
+
+    /**
+     * {@link FtRemote.Script} that does a POST request with a "test" body.
+     * @author Rui Castro (rui.castro@gmail.com)
+     * @version $Id$
+     * @since 0.43
+     */
+    private class FtRemoteScriptPost implements FtRemote.Script {
```
Comment (essobedo) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @rui-castro I don't get it

File: src/test/java/org/takes/tk/TkSlf4jTest.java (hunk 6 above)
Comment (rui-castro) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @essobedo the assert statements are in `FtRemoteScriptPost`.

Comment (essobedo) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @rui-castro one more good reason to make the class anonymous

File: src/main/java/org/takes/tk/TkConsumeBody.java (hunk 2 above)
Comment (rui-castro) [inline src/main/java/org/takes/tk/TkConsumeBody.java]: @essobedo from [here](https://github.com/yegor256/takes/blob/master/pom.xml#L32), but is the wrong version :). What version should I put in `@since` tag?

File: src/test/java/org/takes/tk/TkSlf4jTest.java (hunk 8 above)
Comment (rui-castro) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @essobedo originally it was an anonymous class, but Qulice complained that `TkSlf4jTest` had to much data coupling (it was using too many other classes, 10 or 11, when the maximum was 7). I moved that anonymous class to a different class to reduce that number and pass Qulice control.

File: src/main/java/org/takes/http/BkBasic.java (hunk 1 above)
Comment (rui-castro) [inline src/main/java/org/takes/http/BkBasic.java]: @essobedo do you want me to revert this?

File: src/main/java/org/takes/tk/TkConsumeBody.java (hunk 2 above)
Comment (essobedo) [inline src/main/java/org/takes/tk/TkConsumeBody.java]: @rui-castro this is the version of jcabi parent which is not related, what about `1.1.1`?

Comment (rui-castro) [inline src/main/java/org/takes/tk/TkConsumeBody.java]: @essobedo ok

File: src/main/java/org/takes/http/BkBasic.java (hunk 1 above)
Comment (essobedo) [inline src/main/java/org/takes/http/BkBasic.java]: @rui-castro yes I prefer avoid mixing things.

File: src/test/java/org/takes/tk/TkSlf4jTest.java [hunk 9]
```diff
@@ -55,4 +61,36 @@ public void logsException() throws IOException {
...
+        @Override
+        public void exec(final URI home) throws IOException {
+            new JdkRequest(home)
+                .method("POST")
+                .body().set("test").back()
+                .fetch()
+                .as(RestResponse.class)
+                .assertBody(new IsEqual<>(""))
```
Comment (essobedo) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @rui-castro What about `Matchers.isEmptyOrNullString()`  instead of `new IsEqual<>("")`?

**Comment by rui-castro** [comment]
@essobedo new commit with improvements to all points.

File: src/test/java/org/takes/tk/TkSlf4jTest.java (hunk 6 above)
Comment (essobedo) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @rui-castro The name of your unit test is not correct, it should start with a conjugated verb

File: src/test/java/org/takes/rq/CapInputStreamTest.java [hunk 10]
```diff
@@ -70,15 +69,35 @@ public void closesStream() throws Exception {
...
+        final long skipped = input.skip(2);
+        MatcherAssert.assertThat(skipped, Matchers.equalTo(2L));
+    }
+
+    /**
+     * CapInputStream decreases the available bytes after a skip call.
+     * @throws IOException If some problem inside
+     */
+    @Test
+    public void skipDecreaseAvailableBytes() throws IOException {
```
Comment (essobedo) [inline src/test/java/org/takes/rq/CapInputStreamTest.java]: @rui-castro the name of a test must start with a verb so it should rather be something like "decreasesAvailableBytesOnSkip"

File: src/test/java/org/takes/rq/CapInputStreamTest.java [hunk 11]
```diff
@@ -70,15 +69,35 @@ public void closesStream() throws Exception {
...
+     * @throws IOException If some problem inside
      */
     @Test
-    public void skipsOnStream() throws Exception {
-        final long skip = 25L;
-        final InputStream stream = Mockito.mock(InputStream.class);
-        final CapInputStream wrapper = new CapInputStream(stream, 50L);
-        wrapper.skip(skip);
-        Mockito.verify(stream, Mockito.times(1)).skip(skip);
+    public void skipsOnStream() throws IOException {
```
Comment (essobedo) [inline src/test/java/org/takes/rq/CapInputStreamTest.java]: @rui-castro why don't you keep `skipsOnStream` as it was?

File: src/test/java/org/takes/rq/CapInputStreamTest.java [hunk 12]
```diff
@@ -70,15 +69,35 @@ public void closesStream() throws Exception {
...
+        final byte[] bytes = "skipsOnStream".getBytes();
+        final CapInputStream input = new CapInputStream(
+            new ByteArrayInputStream(bytes),
+            bytes.length
+        );
+        final long skipped = input.skip(2);
+        MatcherAssert.assertThat(skipped, Matchers.equalTo(2L));
```
Comment (essobedo) [inline src/test/java/org/takes/rq/CapInputStreamTest.java]: @rui-castro please remove  the local variables `input` and `skipped` as they are only used once, inline everything into your `MatcherAssert.assertThat`

File: src/test/java/org/takes/rq/CapInputStreamTest.java [hunk 13]
```diff
@@ -70,15 +69,35 @@ public void closesStream() throws Exception {
...
+     */
+    @Test
+    public void skipDecreaseAvailableBytes() throws IOException {
+        final byte[] bytes = "skipDecreaseAvailableBytes".getBytes();
+        final CapInputStream input = new CapInputStream(
+            new ByteArrayInputStream(bytes),
+            bytes.length
+        );
+        final long skipped = input.skip(2);
+        MatcherAssert.assertThat(
```
Comment (essobedo) [inline src/test/java/org/takes/rq/CapInputStreamTest.java]: @rui-castro same remark as above, please inline everything, we have no need of `input` and `skipped` as they are used only once

File: src/test/java/org/takes/tk/TkSlf4jTest.java [hunk 14]
```diff
@@ -55,4 +65,27 @@ public void logsException() throws IOException {
         new TkSlf4j(new TkFailure(new IOException(""))).act(new RqFake());
     }
 
+    /**
+     * {@link TkSlf4j} can output an empty body for {@link TkEmpty}.
+     * @throws IOException if some I/O problem occurred.
+     */
+    @Test
+    public void outputsEmptyResponseBody() throws IOException {
```
Comment (essobedo) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @rui-castro by convention we throw `Exception` in tests

File: src/test/java/org/takes/tk/TkSlf4jTest.java [hunk 15]
```diff
@@ -55,4 +65,27 @@ public void logsException() throws IOException {
         new TkSlf4j(new TkFailure(new IOException(""))).act(new RqFake());
     }
 
+    /**
+     * {@link TkSlf4j} can output an empty body for {@link TkEmpty}.
+     * @throws IOException if some I/O problem occurred.
```
Comment (essobedo) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @rui-castro please remove the trailing dot in the description of your exception

File: src/test/java/org/takes/rq/CapInputStreamTest.java (hunk 11 above)
Comment (rui-castro) [inline src/test/java/org/takes/rq/CapInputStreamTest.java]: @essobedo ok. I will restore the original method.

File: src/test/java/org/takes/rq/CapInputStreamTest.java (hunk 10 above)
Comment (rui-castro) [inline src/test/java/org/takes/rq/CapInputStreamTest.java]: @essobedo ok

File: src/test/java/org/takes/rq/CapInputStreamTest.java (hunk 12 above)
Comment (rui-castro) [inline src/test/java/org/takes/rq/CapInputStreamTest.java]: @essobedo ok

File: src/test/java/org/takes/tk/TkSlf4jTest.java (hunk 14 above)
Comment (rui-castro) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @essobedo ok

File: src/test/java/org/takes/tk/TkSlf4jTest.java (hunk 15 above)
Comment (rui-castro) [inline src/test/java/org/takes/tk/TkSlf4jTest.java]: @essobedo ok

File: src/test/java/org/takes/rq/CapInputStreamTest.java (hunk 13 above)
Comment (essobedo) [inline src/test/java/org/takes/rq/CapInputStreamTest.java]: @rui-castro you are right, ignore this remark then

**Comment by essobedo** [comment]
@rultor merge

File: src/main/java/org/takes/rq/CapInputStream.java (hunk 3 above)
Comment (essobedo) [inline src/main/java/org/takes/rq/CapInputStream.java]: @rui-castro please cast explicitly `this.origin.available()` to `long`

File: src/main/java/org/takes/tk/TkConsumeBody.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Override
+    public Response act(final Request req) throws IOException {
+        final Request reqsafe = new RqChunk(new RqLengthAware(req));
+        final Response rsp = this.origin.act(reqsafe);
+        final InputStream body = reqsafe.body();
+        for (int count = body.available(); count > 0;
+            count = body.available()) {
+            if (body.skip((long) count) < (long) count) {
```
Comment (essobedo) [inline src/main/java/org/takes/tk/TkConsumeBody.java]: @rui-castro why don't you read it explicitly the stream instead of relying on `available()` and `skip`, this way you would not need to modify the current behavior of those methods. You could then create a new ticket to fix those methods, this way we would apply the one PR per fix rule.

Comment (rui-castro) [inline src/main/java/org/takes/tk/TkConsumeBody.java]: @essobedo I will try that and push a new version.

**Comment by rui-castro** [comment]
@essobedo new commit with a simplification.

File: src/main/java/org/takes/tk/TkConsumeBody.java (hunk 16 above)
Comment (rui-castro) [inline src/main/java/org/takes/tk/TkConsumeBody.java]: @essobedo I've tried it to revert all changes to CapInputStrem and just read() the body, but it doesn't work. The read() method gets stuck, because CapInputStream.available() returns available bytes when there aren't any more bytes to read.
The issue is really with the CapInputStream.

File: src/main/java/org/takes/rq/CapInputStream.java (hunk 3 above)
Comment (rui-castro) [inline src/main/java/org/takes/rq/CapInputStream.java]: @essobedo done

**Comment by essobedo** [comment]
@rultor merge

**Comment by rui-castro** [comment]
@yegor256 review this PR, please.

**Comment by essobedo** [comment]
@davvd please proceed here

**Comment by rui-castro** [comment]
@essobedo thank you.

**Comment by essobedo** [comment]
@rui-castro sorry for you, the ticket is obviously not anymore assigned to you, could you please close the PR. Next time don't forget to ask more time to the PM who is davvd in case of this project

**Comment by rui-castro** [comment]
@essobedo I didn't remember I could ask for [more time](http://at.teamed.io/policy.html#10). Thanks for the tip.
The thing that is bothering me, right now, is that the new PR #717 is a copy of mine #716.

**Comment by essobedo** [comment]
@rui-castro yes I know, I'm the reviewer of the other PR

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rui-castro please, close this PR, it's definitely expired. As @essobedo said above, this is how we work, our rules are rather strict. Please, remind me in #712 (when it's closed) to compensate you for this failure, as a one time exception.

**Comment by rui-castro** [comment]
@yegor256 closing this. Thank you.

**Comment by davvd** [comment]
@ypshenychka please, review this task for compliance with our [quality rules](http://at.teamed.io/qa.html)

**Comment by ypshenychka** [comment]
@rui-castro 
According to our [QA Rules](http://at.teamed.io/qa.html):
>Messages in a ticket always start with a name of a user they are addressed to.

Please correct your code review comments by indicating an addressee in the beginning ([here](https://github.com/yegor256/takes/pull/716#discussion_r89613813) and [here](https://github.com/yegor256/takes/pull/716#discussion_r89613822)).

**Comment by rui-castro** [comment]
@ypshenychka done.

**Comment by ypshenychka** [comment]
@rui-castro Thank you!

**Comment by ypshenychka** [comment]
@davvd Quality is acceptable.

**Comment by davvd** [comment]
> @davvd Quality is acceptable.

@ypshenychka thanks for the review, we'll try better next time

**Comment by davvd** [comment]
@essobedo paied 10 mins to @ypshenychka for QA review (payment ID is `111976548`). thanks, paid, **49 mins** to your account, payment ID is `111976584`, task took 603 hours and 20 mins to complete. extra minutes for review comments (c=34). +49 added to your rating, at the moment it is: [+154](http://www.netbout.com/b/36421?open=rating)
